### PR TITLE
Refine wildcard behavior to support prerelease versions

### PR DESCRIPTION
**Pull request by scothis**
Ranges operating on pre release versions can produce counter intuitive results. One might expect the range `>=1.0.0 <2.0.0` to always match versions with major version `1`. However, pre release version are less than the release version, so `1.0.0-build.1` will not match, while `2.0.0-build.1` will. If a range will operate on pre release versions, be sure to include a pre release value of `0` making the desired range `>=1.0.0-0 <2.0.0-0`. Alternatively, the range `1.x` will match all versions with major version `1`.

Without this fix, ranges like `1.x` match `2.0.0-build.1` and fail to match `1.0.0-build.1`.

Note: this changes existing behavior, so we should be cautious. The scope of this change is limited to wildcards and updates behavior that users find surprising. As wildcards (and ranges) are not part of the semver spec, compliance is not affected.

### CODE REVIEW

**Comment by voor** [comment]
Oh this is neat, I definitely had an issue related to this at some point in the past, my silly workaround was instead of doing `>=1.0.0 <2.0.0` I'd just do `>=1.0.0 <1.999.0`

File: README.md [hunk 1]
```diff
@@ -87,6 +87,8 @@ Ranges can be combined by both AND and OR
 
   - `>1.0.0 <2.0.0 || >3.0.0 !4.2.1` would match `1.2.3`, `1.9.9`, `3.1.1`, but not `4.2.1`, `2.1.1`
 
+Ranges operating on pre release versions can produce counter intuitive results. One might expect the range `>=1.0.0 <2.0.0` to always match versions with major version `1`. However, pre release version are less than the release version, so `1.0.0-build.1` will not match, while `2.0.0-build.1` will. If a range will operate on pre release versions, be sure to include a pre release value of `0` making the desired range `>=1.0.0-0 <2.0.0-0`. Alternatively, the range `1.x` will match all versions with major version `1`.
```
Comment (mamachanko) [inline README.md]:
```suggestion
Ranges operating on pre release versions can produce counter intuitive results. One might expect the range `>=1.0.0 <2.0.0` to always match versions with major version `1`. However, pre release versions are less than the release version, so `1.0.0-build.1` will not match, while `2.0.0-build.1` will. If a range will operate on pre release versions, be sure to include a pre release value of `0` making the desired range `>=1.0.0-0 <2.0.0-0`. Alternatively, the range `1.x` will match all versions with major version `1`.
```

**Review by mamachanko — COMMENTED** [review]
LGTM but I do not have an intuition concerning its impact.

At this moment the following Carvel tools use this fork of `semver`:
* `vendir`: https://github.com/carvel-dev/vendir/blob/develop/go.mod
* `kbld` (indirectly through `vendir`): https://github.com/carvel-dev/kbld/blob/develop/go.mod
* `kapp-controller` (indirectly through `vendir`): https://github.com/carvel-dev/kapp-controller/blob/develop/go.mod

**Comment by neil-hickey** [comment]
hey @scothis I am gonna check what other think for this topic. I can't say I am a semver purist. But as you pointed out, semantically 1.0.0-rc.1 is actually before 1.0.0. So the range is actually what I would personally expect right now, be it intuitive or not - it is 'correct'. So I am a little uneasy on making this change. Happy to hear from others though.

Just getting a sense from other libraries:

npm's semver library: https://www.npmjs.com/package/semver

Has the following behaviour:

```
const semver = require('semver')

console.log(semver.satisfies('1.0.0-build.1', '>1.0.0-0 <3.0.0-0')) // true
```

```
const semver = require('semver')

console.log(semver.satisfies('2.0.0-build.1', '>1.0.0-0 <3.0.0-0')) // false
```

Reason for false: `Second, a user who has opted into using a prerelease version has clearly indicated the intent to use that specific set of alpha/beta/rc versions. By including a prerelease tag in the range, the user is indicating that they are aware of the risk. However, it is still not appropriate to assume that they have opted into taking a similar risk on the next set of prerelease versions.`

```
const semver = require('semver')

console.log(semver.satisfies('3.0.0-build.1', '>1.0.0-0 <3.0.0-0')) // false
```

**Comment by scothis** [comment]
@neil-hickey to clarify, there is no change in behavior to this library relating to the range queries you showed using the npm server module. The strict ordering of versions and range comparisons using less than or greater than operators remain unchanged. The carvel tools have criteria other than the version range for whether prerelease versions should be included, or to further filter which prerelease version should be considered, those are also unchanged.

The only behavior change I'm proposing in this PR is for a constraint like `2.x` to match `2.0.0-build.1` and not match `3.0.0-build.1`. This is not the behavior today.

The upstream library had no test coverage for ranges with prerelease version, but always considers them in the evaluation. I believe the current behavior to be an oversight rather than design intent.

**Pull request merged**